modules/ldap/src/main/java/com/neighborhoods/wasabi/ldap/impl/LdapDelegate.java

```
@@ -0,0 +1,857 @@
+            if (attr != null) {
+                email = attr.getString();
+            }
+            return new DirectoryUser(new UserInfo.Builder(UserInfo.Username.valueOf(username)).withFirstName(firstName)
+                    .withLastName(lastName).withEmail(email).withUserId(username).build());
+        }
+        return null;
+    }
+
+    /**
+     * Gets the ldap connection.
+     * 
+     * @return the ldap connection
+     */
+    protected LdapConnection getLdapConnection() {
+        return new LdapNetworkConnection(this.config.getLdapHost(), this.config.getLdapPort(),
+                this.config.isSecureConnection());
+    }
+
+    /**
+     * Gets a generic LDAP info connection
+     *
+     * @return the ldap info connection
+     * @throws LdapException the ldap exception
+     */
+    protected LdapConnection getLdapInfoConnection() throws LdapException {
+        LdapConnection connection = getLdapConnection();
+        connection.bind(this.config.getInfoDN(), this.config.getInfoPassword());
+        return connection;
+    }
+
+    /**
+     * Gets the cached user search request (and builds it if necessary).
+     *
+     * @return the user cache search request
+     * @throws LdapException If there are DN validation errors or invalid filters
+     */
+    protected SearchRequest getUserCacheSearchRequest() throws LdapException {
+        // This should never be null unless it is the first request
+        if (userCacheSearchRequest == null) {
+            // In unlikely event multiple threads reach this point, lock initialization
+            synchronized (this) {
+                SearchRequest searchRequest = new SearchRequestImpl();
+                searchRequest.setScope(SearchScope.SUBTREE);
+                searchRequest.addAttributes(this.config.getRoleAttribute());
+                searchRequest.addAttributes(this.config.getMembershipAttribute());
+                searchRequest.setTimeLimit(0);
```

do we not want to timeout eventually if there's a connection issue? we can throw an exception or something.

---

This is actually after we already have the connection and perform a search. The connection interrupted exception would be thrown if we lost connectivity to the server. 

Since this *could* be a very long running search for a large organization with many users, I wasn't sure what an appropriate limit might be since it varies based on their LDAP contents.

But sure, I've set a limit of a max search time of 5 minutes. Do you think that is reasonable?

---

5 minutes sounds reasonable to me.

---

yep